databuilder/extractor/bigquery_usage_extractor.py

```diff
@@ -0,0 +1,103 @@
+from collections import namedtuple
+from datetime import date, timedelta
+import logging
+
+from googleapiclient.discovery import build
+from oauth2client.client import GoogleCredentials
+
+from databuilder.extractor.base_extractor import Extractor
+
+TableColumnUsageTuple = namedtuple('TableColumnUsageTuple', ['database', 'cluster', 'schema',
+                                                             'table', 'column', 'email'])
+
+LOGGER = logging.getLogger(__name__)
+
+
+class BigQueryUsageExtractor(Extractor):
+    """
+    An aggregate extractor for bigquery table usage. This class takes the data from
+    the stackdriver logging API by filtering on timestamp, bigquery_resource and looking
+    for referencedTables in the response.
+    """
+    TIMESTAMP_KEY = 'timestamp'
+    PROJECT_ID_KEY = 'projectid'
+
+    def init(self, conf):
+        credentials = GoogleCredentials.get_application_default()
+        self.logging_service = build('logging', 'v2', credentials=credentials)
+
+        # type: (ConfigTree) -> None
+        self.timestamp = conf.get_string(
+            BigQueryUsageExtractor.TIMESTAMP_KEY,
+            (date.today() - timedelta(days=1)).strftime('%Y-%m-%dT00:00:00Z'))
+        self.projectid = conf.get_string(BigQueryUsageExtractor.PROJECT_ID_KEY)
+        self.table_usage_counts = {}
+        self._count_usage()
+        self.iter = iter(self.table_usage_counts)
+
+    def _count_usage(self):
+        for entry in self._retrieve_records():
+            if not isinstance(entry, dict):
+                raise Exception("BigQueryUsageExtractor expects record of type dict!")
+
+            email = entry['protoPayload']['authenticationInfo']['principalEmail']
+            jobCompletedEvent = entry['protoPayload']['serviceData']['jobCompletedEvent']
+            refTables = jobCompletedEvent['job']['jobStatistics'].get('referencedTables', None)
```

Comment:
will the key specified in  the long chain dictionary be missing in some cases?

Reply:
Yes, if the exact same query is processed again, the BQ results will be read from a cached table, so the source tables are not actually touched. I have observed that then the referencedTables will not be populated. The cached table is available for 24 hours or until any of the source tables are touched.

Reply:
Could we link BQ documentation on this access log data model as a code comment here for easy reference?

Reply:
The docs do not explain this, but I observed it. I did find another entry in the docs for "totalTablesProcessed". From what I interpret from the docs, historically only the first N tables were listed, but the docs imply as if this was expanded to the full list recently.

Reply:
So what I did was to look at that and see if len(refTables) matches the totalTablesProcessed and otherwise write a warning in the log.